File: app/src/main/java/com/starostinvlad/fan/VideoScreen/PlayerModel/SecondPlayerDataDeserializer.java
package com.starostinvlad.fan.VideoScreen.PlayerModel;

import android.annotation.SuppressLint;
import android.util.Log;

import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map.Entry;

public class SecondPlayerDataDeserializer implements JsonDeserializer<Serial> {
    private final String TAG = getClass().getSimpleName();
    private final String host;
    private int playerId;
    private int seasonNumber = 0;

    SecondPlayerDataDeserializer(int id, String host) {
        this.playerId = id;
        this.host = host;
    }

    // TODO: 12.04.2021 распределить работу на методы

    @Override
    public Serial deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
//        Log.d(TAG, "deserialize2: start");
        Serial serial = new Serial();
        JsonObject jsonObject = json.getAsJsonObject();
//        Log.d(TAG, "deserialize2: " + jsonObject.keySet().toString());
        List<Season> seasons = new ArrayList<>();
        for (Entry<String, JsonElement> jsonSeason : jsonObject.entrySet()) {
            seasonNumber = Integer.parseInt(jsonSeason.getKey());
            List<Episode> episodes = new ArrayList<>();
            if (jsonSeason.getValue().isJsonObject()) {
                JsonObject seasonObject = jsonSeason.getValue().getAsJsonObject();

                for (Entry<String, JsonElement> jsonEpisode : seasonObject.entrySet()) {
                    int episodeNumber = Integer.parseInt(jsonEpisode.getKey());
                    JsonObject episodeObject = jsonEpisode.getValue().getAsJsonObject();
                    episodes.add(
                            createEpisode(episodeNumber, episodeObject)
                    );
                }
            } else {
                JsonArray episodeJsonArray = jsonSeason.getValue().getAsJsonArray();
                int episodeNumber = 0;
//                Log.d(TAG, "deserialize: episode: " + episodeNumber);
                for (JsonElement episodeJsonElement : episodeJsonArray) {
                    episodes.add(
                            createEpisode(episodeNumber, episodeJsonElement.getAsJsonObject())
                    );
                    episodeNumber++;
                }
            }
            Season season = new Season();
            season.setNumber(seasonNumber);
            season.setEpisodes(episodes);

            seasons.add(season);
        }
        serial.setSeasonList(seasons);
//        Log.d(TAG, "deserialize: end! tsize: " + translations.size());
        return serial;
    }

    @SuppressLint("DefaultLocale")
    private Episode createEpisode(int episodeNumber, JsonObject jsonObject) {
        Episode episode = new Episode();
        episode.setNumber(episodeNumber);
        episode.setTitle(
                String
                        .format(
                                "%d сезон %s серия",
                                seasonNumber,
                                (episodeNumber != 0 ? episodeNumber : "спецвыпуск")
                        )
        );
        episode.setTranslations(
                parseTranslations(
                        jsonObject,
                        seasonNumber,
                        episodeNumber
                )
        );
        return episode;
    }

    @SuppressLint("DefaultLocale")
    private List<Translation> parseTranslations(JsonObject episodeObject, int season, int episodeNumber) {
        List<Translation> translations = new ArrayList<>();
        Log.d(TAG, "parseTranslations: jsonOBject: " + episodeObject.toString());
        for (Entry<String, JsonElement> elementEntry : episodeObject.entrySet()) {
            Translation translation = new Translation();

            String[] voiceString = elementEntry.getKey().split("#");
            int translationCode = Integer.parseInt(voiceString[0]);
            String voiceTitle = voiceString[1];
            int uniqueId = elementEntry.getValue().getAsInt();
            //https://sorb.info/dew/3949?season=1&episode=1&voice=1&alloff=true
            //"https://%s/player/responce.php?id=%d&season=%d&episode=%d&voice=%d&uniqueid=%d",
            String url = String.format(
//                    "https://%s/dew/%d?season=%d&episode=%d&voice=%d&alloff=true&uniqueid=%d",
                    "https://%s/player/responce.php?id=%d&season=%d&episode=%d&voice=%d&uniqueid=%d",
                    host,
                    playerId,
                    season,
                    episodeNumber,
                    translationCode,
                    uniqueId
            );
            translation.setTitle(voiceTitle);
            translation.setId(uniqueId);
            translation.setCode(translationCode);
            translation.setUrl(url);
            translations.add(translation);
        }
        return translations;
    }
}

File: app/src/main/java/com/starostinvlad/fan/VideoScreen/PlayerModel/SecondPlayerDataDeserializer.kt
package com.starostinvlad.fan.VideoScreen.PlayerModel

import android.annotation.SuppressLint
import android.util.Log
import com.google.gson.*
import java.lang.reflect.Type
import java.util.*
import kotlin.Throws

class SecondPlayerDataDeserializer internal constructor(private val playerId: Int, private val host: String?) : JsonDeserializer<Serial> {
    private val TAG: String = javaClass.getSimpleName()
    private var seasonNumber = 0

    // TODO: 12.04.2021 распределить работу на методы
    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Serial {
//        Log.d(TAG, "deserialize2: start");
        val serial = Serial()
        val jsonObject = json.asJsonObject
        //        Log.d(TAG, "deserialize2: " + jsonObject.keySet().toString());
        val seasons: MutableList<Season> = ArrayList()
        for ((key, value) in jsonObject.entrySet()) {
            seasonNumber = key.toInt()
            val episodes: MutableList<Episode> = ArrayList()
            if (value.isJsonObject) {
                val seasonObject = value.asJsonObject
                for ((key1, value1) in seasonObject.entrySet()) {
                    val episodeNumber: Int = key1.toInt()
                    val episodeObject = value1.asJsonObject
                    episodes.add(
                            createEpisode(episodeNumber, episodeObject)
                    )
                }
            } else {
                val episodeJsonArray = value.asJsonArray
                var episodeNumber = 0
                //                Log.d(TAG, "deserialize: episode: " + episodeNumber);
                for (episodeJsonElement in episodeJsonArray) {
                    episodes.add(
                            createEpisode(episodeNumber, episodeJsonElement.asJsonObject)
                    )
                    episodeNumber++
                }
            }
            val season = Season()
            season.number = seasonNumber
            season.episodes = episodes
            seasons.add(season)
        }
        serial.seasonList = seasons
        //        Log.d(TAG, "deserialize: end! tsize: " + translations.size());
        return serial
    }

    @SuppressLint("DefaultLocale")
    private fun createEpisode(episodeNumber: Int, jsonObject: JsonObject): Episode {
        val episode = Episode()
        episode.number = episodeNumber
        episode.title = String.format(
                "%d сезон %s серия",
                seasonNumber,
                if (episodeNumber != 0) episodeNumber else "спецвыпуск"
        )
        episode.translations = parseTranslations(
                jsonObject,
                seasonNumber,
                episodeNumber
        )
        return episode
    }

    @SuppressLint("DefaultLocale")
    private fun parseTranslations(episodeObject: JsonObject, season: Int, episodeNumber: Int): MutableList<Translation> {
        val translations: MutableList<Translation> = ArrayList()
        Log.d(TAG, "parseTranslations: jsonOBject: $episodeObject")
        for ((key, value) in episodeObject.entrySet()) {
            val translation = Translation()
            val voiceString: Array<String> = key.split("#".toRegex()).toTypedArray()
            val translationCode: Int = voiceString[0].toInt()
            val voiceTitle = voiceString[1]
            val uniqueId = value.asInt
            //https://sorb.info/dew/3949?season=1&episode=1&voice=1&alloff=true
            //"https://%s/player/responce.php?id=%d&season=%d&episode=%d&voice=%d&uniqueid=%d",
            val url: String = String.format( //                    "https://%s/dew/%d?season=%d&episode=%d&voice=%d&alloff=true&uniqueid=%d",
                    "https://%s/player/responce.php?id=%d&season=%d&episode=%d&voice=%d&uniqueid=%d",
                    host,
                    playerId,
                    season,
                    episodeNumber,
                    translationCode,
                    uniqueId
            )
            translation.title = voiceTitle
            translation.id = uniqueId
            translation.code = translationCode
            translation.url = url
            translations.add(translation)
        }
        return translations
    }
}